utilities/object_registry.cc

```diff
@@ -12,49 +12,100 @@
 
 namespace ROCKSDB_NAMESPACE {
 #ifndef ROCKSDB_LITE
-// Looks through the "type" factories for one that matches "name".
-// If found, returns the pointer to the Entry matching this name.
-// Otherwise, nullptr is returned
-const ObjectLibrary::Entry *ObjectLibrary::FindEntry(
-    const std::string &type, const std::string &name) const {
-  std::unique_lock<std::mutex> lock(mu_);
-  auto entries = entries_.find(type);
-  if (entries != entries_.end()) {
-    for (const auto &entry : entries->second) {
-      if (entry->matches(name)) {
-        return entry.get();
+bool ObjectLibrary::PatternEntry::MatchesPattern(const std::string &name,
+                                                 size_t nlen,
+                                                 const std::string &target,
+                                                 size_t tlen) const {
+  if (nlen == tlen) {  // The lengths are the same
+    return (name_only_ || patterns_.empty()) && name == target;
+  } else if (patterns_.empty()) {
+    return nlen == tlen && name == target;
```

Comment:
the nlen==tlen at line 22 is meant to prevent the comparison of strings on the second half.  If the nlen==tlen, it will be caught at line 19 but if they are not equal then the check here could speed up the check.

I re-ordered the cases to simplify the code slightly.